src/share/poudriere/common.sh

```diff
@@ -7729,31 +7729,45 @@ calculate_size_in_mb() {
 }
 
 calculate_ospart_size() {
+	local NUM_PART
 	local CALC_SIZE
 	local FULL_CALC_SIZE
 	local DATA_CALC_SIZE
+	local SWAP_CALC_SIZE
 	local CFG_CALC_SIZE
 
+	# How many partitions do we need
+	NUM_PART=$1
+
 	# Figure out the size of the image in MB
-	CALC_SIZE=${IMAGESIZE}
+	CALC_SIZE=$2
 	calculate_size_in_mb
 	FULL_CALC_SIZE=${CALC_SIZE}
 
 	# Figure out the size of the /cfg partition
-	CALC_SIZE=${CFG_SIZE}
+	CALC_SIZE=$3
 	calculate_size_in_mb
 	CFG_CALC_SIZE=${CALC_SIZE}
 
 	# Figure out the size of the Data partition
-	if [ $# -eq 3 -o -n ${DATA_SIZE} ]; then
-		CALC_SIZE=${DATA_SIZE}
+	if [ $# -ge 4 -o -n $4 ]; then
```

Comment:
I think this pattern can be error prone and it is best to resolve all params up front like:
```
local NUM_PART="$1"
local IMAGESIZE="$2"
local CFG_SIZE="$3"
local DATA_SIZE="$4"
local SWAP_CALC_SIZE="$5"
```
...
```
if [ -n "$DATA_SIZE" ]; then
CALC_SIZE="$DATA_SIZE"
...
fi
```
...
```
if [ -n "$SWAP_CALC_SIZE" ]; then
CALC_SIZE="$SWAP_CALC_SIZE"
...
fi
```

Also having said all of that I don't know why `CALC_SIZE` is a thing. `calculate_size_in_mb` should take a proper parameter. It can return the value wanted directly.
```
calculate_size_in_mb() {                       
        local CALC_SIZE="$1"
        local var_return="$2"

        case ${CALC_SIZE} in                   
        *p)                                    
                CALC_SIZE=${CALC_SIZE%p}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *t)                                    
                CALC_SIZE=${CALC_SIZE%t}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *g)                                    
                CALC_SIZE=${CALC_SIZE%g}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *m)                                    
                CALC_SIZE=${CALC_SIZE%m}       
        esac 
       
        setvar "$var_return" "$CALC_SIZE"                                  
}                                              
```

Then called like:
```
calculate_size_in_mb "$SWAP_CALC_SIZE" SWAP_CALC_SIZE
```
It would be easy to make it `calculate_size_in_mb SWAP_CALC_SIZE` too if that is more appealing:
```
calculate_size_in_mb() {                       
        local calc_var="$1"
        local CALC_SIZE

        getvar "$calc_var" CALC_SIZE

        case ${CALC_SIZE} in                   
        *p)                                    
                CALC_SIZE=${CALC_SIZE%p}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *t)                                    
                CALC_SIZE=${CALC_SIZE%t}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *g)                                    
                CALC_SIZE=${CALC_SIZE%g}       
                CALC_SIZE=$((CALC_SIZE << 10 ))
                ;&                             
        *m)                                    
                CALC_SIZE=${CALC_SIZE%m}       
        esac 
       
        setvar "$calc_var" "$CALC_SIZE"                                  
}  
```
...
```
calculate_size_in_mb SWAP_CALC_SIZE
```

Reply:
For reference the `getvar` is just a wrapper around an `eval` but `setvar` is a FreeBSD sh extension.

Reply:
In case my edit is lost in the noise my last `calculate_size_in_mb` has a bug.
```
### Can't use CALC_SIZE in this function for it to work or the setvar just sets the
### local version of the variable. I didn't have a real editor for this but a simple
### s/CALC_SIZE/calc_size/ in the function would avoid the problem.
```

Reply:
Yeah, that works out much better